### DISCUSSION TITLE
`#[staticmethod]` in subclass

### DISCUSSION THREAD

**Discussion by Rexagon** [PyO3/pyo3#2978] [Questions] [answered]
This library is amazing in terms of usability, api is very intuitive!

However I stumbled upon an unexpected problem. Consider the following code:
```rust
#[pyclass(subclass)]
pub struct Seed(...);

#[pyclass(extends = Seed)]
pub struct Bip39Seed;

#[pymethods]
impl Bip39Seed {
    #[staticmethod]
    fn generate() -> PyClassInitializer<Self> {
        ...
        PyClassInitializer::from(seed).add_subclass(Self)
    }
}
``` 

There I'm trying to create a subclass with a static method. However, for the result type it expects something to be casted to `PyObject` and is unable to compile `PyClassInitializer`:
```
the trait `pyo3::IntoPy<pyo3::Py<pyo3::PyAny>>` is not implemented for `pyo3::PyClassInitializer<crypto::Bip39Seed>`
```

Changing result type to `(Self, Seed)` doesn't work either.

What is a recommended way to resolve this issue?

(It seems that [`PyObjectInit`](https://docs.rs/pyo3/0.18.1/pyo3/pyclass_init/trait.PyObjectInit.html) can be used, but I hope it is possible to avoid unsafe for simple things)

**Comment by davidhewitt** [comment, marked as answer]
Try

```rust
    #[staticmethod]
    fn generate(py: Python) -> PyResult<Py<Self>> {
        ...
        Py::new(py, PyClassInitializer::from(seed).add_subclass(Self))
    }
```
